### PR TITLE
[utilities] Create utility classes for interacting with flex counters

### PR DESCRIPTION
**Pull request by daall**
[utilities] Create utility classes for interacting with flex counters
- Adds utility classes for interacting with flex counters
- Updates makefiles to support new utilities

Signed-off-by: Danny Allen <daall@microsoft.com>

**Details if related**
Related to/spun off from: Azure/sonic-swss#1075

### CODE REVIEW

**Comment by lguohan** [comment]
are you planning to leverage this flex counter manager class for other places using flex counter?

**Comment by daall** [comment]
> are you planning to leverage this flex counter manager class for other places using flex counter?

That's my goal, yes. Eventually I would want to refactor other daemons like the pfcorch and portsorch to use FlexCounterManager instead of re-implementing the same redis messaging schemes. To do that I'll have to extend the functionality to support plugins and such, but I wanted to limit the scope to features needed for the debug counters for this initial PR.

**Comment by daall** [comment]
retest this please

File: orchagent/flex_counter/dynamic_flex_counter_manager.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        const sai_object_id_t object_id,
+        const CounterType counter_type,
+        const string &counter_stat)
+{
+    SWSS_LOG_ENTER();
+
+    auto counter_stats = object_stats.find(object_id);
+    if (counter_stats == object_stats.end()) 
+    {
+        SWSS_LOG_DEBUG("Could not find flex stat '%s' on object '%s'",
```
Comment (lguohan) [inline orchagent/flex_counter/dynamic_flex_counter_manager.cpp]: is this warning level? such event should not happen in normal operation.

File: orchagent/flex_counter/flex_counter_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+// serializeCounterStats turns a set of stats into a format suitable for FLEX_COUNTER_DB.
+string FlexCounterManager::serializeCounterStats(
+        const unordered_set<string> &counter_stats) const
+{
+    SWSS_LOG_ENTER();
+
+    string stats_string;
+    for (const auto &stat : counter_stats) 
+    {
+        stats_string.append(",");
```
Comment (qiluo-msft) [inline orchagent/flex_counter/flex_counter_manager.cpp]: >stats_string.append(","); [](start = 8, length = 25)

Don't append before first stat.

File: orchagent/flex_counter/dynamic_flex_counter_manager.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        const CounterType counter_type,
+        const string &counter_stat)
+{
+    SWSS_LOG_ENTER();
+
+    auto counter_stats = object_stats.find(object_id);
+    if (counter_stats == object_stats.end())
+    {
+        unordered_set<string> new_stats = { counter_stat };
+        object_stats.emplace(std::make_pair(object_id, new_stats));
```
Comment (qiluo-msft) [inline orchagent/flex_counter/dynamic_flex_counter_manager.cpp]: >std::make_pair [](start = 29, length = 14)

No need to make_pair

File: orchagent/flex_counter/dynamic_flex_counter_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        counter_stats = object_stats.find(object_id);
```
Comment (qiluo-msft) [inline orchagent/flex_counter/dynamic_flex_counter_manager.cpp]: >find [](start = 37, length = 4)

No need to `find`, you get an iterator when `emplace`.

File: orchagent/flex_counter/dynamic_flex_counter_manager.h [hunk 5]
```diff
@@ -0,0 +1,33 @@
+#ifndef SWSS_UTIL_DYNAMIC_FLEX_COUNTER_MANAGER_H
+#define SWSS_UTIL_DYNAMIC_FLEX_COUNTER_MANAGER_H
+
+#include "flex_counter_manager.h"
+
+// DynamicFlexCounterManager allows users to manage a group of flex counters
+// where the objects have highly variable sets of stats to track.
+class DynamicFlexCounterManager : public FlexCounterManager
```
Comment (qiluo-msft) [inline orchagent/flex_counter/dynamic_flex_counter_manager.h]: >DynamicFlexCounterManager [](start = 6, length = 25)

Name is confusing here. Is it dynamic counter or dynamic manager? For my reading, I guess it is `FlexCounterStatManager`

Comment (qiluo-msft) [reply orchagent/flex_counter/dynamic_flex_counter_manager.h]: Another option is to combine FlexCounterManager with DynamicFlexCounterManager into one class.

---
In reply to: [340903276](https://github.com/Azure/sonic-swss/pull/1093#discussion_r340903276) [](ancestors = 340903276)

File: orchagent/flex_counter/flex_counter_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            group_name.c_str());
+}
+
+string FlexCounterManager::getFlexCounterTableKey(
+        const string &group_name, 
+        const sai_object_id_t object_id) const
+{
+    SWSS_LOG_ENTER();
+
+    return group_name + ':' + sai_serialize_object_id(object_id);
```
Comment (lguohan) [inline orchagent/flex_counter/flex_counter_manager.cpp]: can you use TableBase::tableNameSeparatorMap:COUNTERS_DB?

File: orchagent/flex_counter/dynamic_flex_counter_manager.h (hunk 5 above)
Comment (daall) [reply orchagent/flex_counter/dynamic_flex_counter_manager.h]: I'll go with your FlexCounterStatManager suggestion! The reason they're separate is related to this discussion: https://github.com/Azure/sonic-swss/pull/1093#issuecomment-548103642. Most of the existing daemons only set the stats once, so there's no need to maintain a mapping with all the objects and their stats.

File: orchagent/flex_counter/flex_counter_manager.cpp (hunk 2 above)
Comment (daall) [reply orchagent/flex_counter/flex_counter_manager.cpp]: This is actually intentional but may have been unclear. I've added a comment to clarify the purpose.

Comment (qiluo-msft) [reply orchagent/flex_counter/flex_counter_manager.cpp]: I know your intention. I mean don't append before first component, not a good practice.

**Comment by daall** [comment]
retest this please

**Comment by yxieca** [comment]
Noticed some coding style issue. I think the reference type is usually declared in the format of "type& var" instead of "type &var", especially when var is the only variable defined.

Am I too old school? :-)

File: orchagent/flex_counter/flex_counter_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    flex_counter_group_table->set(group_name, field_values);
+
+    SWSS_LOG_DEBUG("Initialized flex counter group '%s'.", group_name.c_str());
+}
+
+FlexCounterManager::~FlexCounterManager()
+{
+    SWSS_LOG_ENTER();
+
+    flex_counter_group_table->del(group_name);
```
Comment (yxieca) [inline orchagent/flex_counter/flex_counter_manager.cpp]: Logically, I think the above line should be after the for loop.

**Comment by daall** [comment]
retest vs please

**Comment by daall** [comment]
retest this please

**Comment by daall** [comment]
retest this please

**Comment by daall** [comment]
retest this please

**Pull request merged**